assembly/src/parser/error.rs

```diff
@@ -0,0 +1,332 @@
+    ImmediateOutOfRange {
+        #[label]
+        span: SourceSpan,
+        range: Range<usize>,
+    },
+    #[error("too many procedures in this module")]
+    #[diagnostic()]
+    ModuleTooLarge {
+        #[label]
+        span: SourceSpan,
+    },
+    #[error("too many re-exported procedures in this module")]
+    #[diagnostic()]
+    ModuleTooManyReexports {
+        #[label]
+        span: SourceSpan,
+    },
+    #[error("too many operands for `push`: tried to push {} elements, but only 16 can be pushed at one time", count)]
+    #[diagnostic()]
+    PushOverflow {
+        #[label]
+        span: SourceSpan,
+        count: usize,
+    },
+    #[error("expected a fully-qualified module path, e.g. `std::u64`")]
+    UnqualifiedImport {
+        #[label]
+        span: SourceSpan,
+    },
+}
+
+impl ParsingError {
+    fn tag(&self) -> u8 {
+        // SAFETY: This is safe because we have given this enum a
+        // primitive representation with #[repr(u8)], with the first
+        // field of the underlying union-of-structs the discriminant
+        unsafe { *<*const _>::from(self).cast::<u8>() }
+    }
+}
+
+impl Eq for ParsingError {}
+
+impl PartialEq for ParsingError {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Failed, Self::Failed) => true,
+            (Self::InvalidLiteral { kind: l, .. }, Self::InvalidLiteral { kind: r, .. }) => l == r,
+            (Self::InvalidHexLiteral { kind: l, .. }, Self::InvalidHexLiteral { kind: r, .. }) => {
+                l == r
+            }
+            (
+                Self::InvalidLibraryPath { message: l, .. },
+                Self::InvalidLibraryPath { message: r, .. },
+            ) => l == r,
+            (
+                Self::ImmediateOutOfRange { range: l, .. },
+                Self::ImmediateOutOfRange { range: r, .. },
+            ) => l == r,
+            (Self::PushOverflow { count: l, .. }, Self::PushOverflow { count: r, .. }) => l == r,
+            (
+                Self::UnrecognizedToken {
+                    token: ltok,
+                    expected: lexpect,
+                    ..
+                },
+                Self::UnrecognizedToken {
+                    token: rtok,
+                    expected: rexpect,
+                    ..
+                },
+            ) => ltok == rtok && lexpect == rexpect,
+            (Self::ExtraToken { token: ltok, .. }, Self::ExtraToken { token: rtok, .. }) => {
+                ltok == rtok
+            }
+            (
+                Self::UnrecognizedEof {
+                    expected: lexpect, ..
+                },
+                Self::UnrecognizedEof {
+                    expected: rexpect, ..
+                },
+            ) => lexpect == rexpect,
+            (x, y) => x.tag() == y.tag(),
+        }
+    }
+}
+
+pub fn handle_parse_error(err: ParseError) -> ParsingError {
```

Comment:
I think we should be able to replace this with implementation of `From<ParseError<'_>> for ParsingError`. I think it would be a bit cleaner this way.

Reply:
True, originally it took additional arguments, but those went away, so we can make this use a conversion trait instead.